### DISCUSSION TITLE
zlib with level=1 available?

### DISCUSSION THREAD

**Discussion by bastie** [tsolomko/SWCompression#32] [Q&A] [answered]
Hello,

how can I compress like ``lib-flate -compress=1 <inputFile >outputFile``?

I need exactly this compression algorithm to be compatible with external specification / programs.

thx
Sebastian

**Comment by tsolomko** [comment, marked as answer]
Sorry for the long response time.

Zlib (or actually Deflate) compression levels are not standardized, see e.g. [Zlib Specification](https://datatracker.ietf.org/doc/html/rfc1950). This is mostly because compression, compared to decompression, is more of an art. You can employ many different techniques and heuristics to get different tradeoffs between compression ratio and time. This is even true for "simpler" algorithms such Deflate/Zlib, as you can choose how you search for subdata matches, do you use static Huffman trees or dynamic ones, how do you construct the dynamic trees, etc. 

As such, compression levels remain an implementation detail and different implementations may define the exact specifics of them in different ways. For example, I do not know what does level 1 of `lib-flate` from your example mean. SWCompression at the moment has only one level, which it _formally_ defines as level 3 (aka "slowest algorithm").

Note, however, that the output _format_ does not depend on the level and that every specification-compliant decompressor must be able to decompress data with any specified level of compression. If you observe, that the output produced by SWCompression is rejected by other implementations as malformed, please report this as an issue, because this may warrant a further investigation.

To answer your direct question, to perform Zlib compression, do the following:
```swift
let data = ... // data that you want to compress
let compressedData = ZlibArchive.archive(data: data)
```
As I mentioned above, this produces data that declares that it was formally compressed with level 3. If this is an issue, a hacky workaround is the following:
```swift
let data = ... // data that you want to compress
var compressedData = ZlibArchive.archive(data: data)
compressedData[1] = 0x5E
```
The last line manually changes the declared compression level to 1.

P.S. A simpler explanation: a compression level is not observable from the compressed data alone, it can only be detected from a certain metadata field, which in fact can be set arbitrarily (from 0 to 3).

**Reply by bastie** [reply]
Thank you for this explanation. At the moment I looking for a swift implementation with same output as git it do. I tested some implementation and the Apple Compression impl. is too good. ``zlib-flate`` (I saw auto correction removed the z) with parameters to compress as "level=1" do this.

with regards
